002-tidyr-stringr.Rmd

```diff
@@ -0,0 +1,133 @@
+
+-   Neither stringr nor tidyr provide a particularly nice way to work with named capture groups.
+    There's also scope for more directly matching column names to patterns if `extract()` didn't attempt to match the syntax of `separate()` so closely, e.g. `str_extract(x, c(type = ".*", "|", fruit = "orange|apple"))`.
+
+-   tidyr only provides one function for splitting a string across rows, rather than columns: `separate_rows()`.
+    What other functions should exist in this family?
+
+-   `tidyr::separate()` can split up a string by position or by pattern.
+    This now seems too clever, and makes it hard to notice that there are two different tools here.
+
+### Components
+
+To find a solution to this problem we need to first enumerate all the parts.
+I think we can break the problem down into two somewhat independent components: how you break the string into pieces, and whether the number of pieces is fixed (and named) or varying (and named).
+In total, I think this gives five ways to split up a string:
+
+-   Into a fixed number of pieces by length/position, e.g. `"1m15"` -> `c(trt = "1", gender = "m", age = "15")`.
+    This is closely related to `readr::read_fwf()`.
+
+-   Into a varying number of pieces by length, e.g. `"abcdef"` -> `c("ab", "cd", "ef")`
+
+-   Into a fixed number of pieces by a pattern, e.g. `"a/b/c"` -> `c(treatment = "a", group = "b", subgroup = "c").`
+
+-   Into a varying number of pieces by a pattern, e.g. `"a/b/c"` -> `c("a", "b", "c").`
+
+-   Into named groups, e.g. `"a(b)c"` -> `c(treatment = "a", group = "b", subgroup = "c")`
+
+Matching these operations up with the existing stringr and tidyr function gives the following table:
+
+| Split by | Length  | stringr         | tidyr             |
+|----------|---------|-----------------|-------------------|
+| Length   | Fixed   |                 | `separate()`      |
+| Length   | Varying |                 |                   |
+| Pattern  | Fixed   | `str_split_n()` | `separate()`      |
+| Pattern  | Varying | `str_split()`   | `separate_rows()` |
+| Group    | Fixed   | `str_match()`   | `extract()`       |
+
+Note that `separate()` is repeated twice, and there are a number of empty cells.
+Interestingly, there's no stringr function that allows you to extract multiple pieces by position; you currently have to call `str_sub()` multiple times yourself.
+
+Generally, when the pieces might vary from entry to entry you'll want to create rows; when the pieces are fixed, you'll want to create columns.
+This might not be true in every case, but I think it's true in enough cases that we can assume it as a constraint for this problem.
+(Please let me know if you have a counter-example.)
+
+We don't need to consider compound problems that need to be solved by splitting into rows and columns simultaneously because you a compound problem (e.g. `"a-b-c/d-e-f/g-h-i"`) can always be solved it by first splitting into rows and then into columns.
```

Comment:
```suggestion
We don't need to consider compound problems that need to be solved by splitting into rows and columns simultaneously because a compound problem (e.g. `"a-b-c/d-e-f/g-h-i"`) can always be solved by first splitting into rows and then into columns.
```